Implement reference mapping components

VladimirShitov opened the pull request:
- Implement `integrate/scarches` component to map query data to the reference (currently only supports HLCA)
- Implement `integrate/scvi_tools` component to build reference and map query data on it
- Add scVI model support for `integrate/scvi_tools`
- Add scanVI model support for `integrate/scvi_tools`
- Add totalVI model support for `integrate/scvi_tools`
- Add scripts for downloading test files for reference mapping

LuckyMD commented:
PR concept feedback:
User workflow 1: User has new lung data and wants to map to known reference -> HLCA
Requirements:
* QC'ed data
* Note: HLCA ref model loaded in code

User workflow 2: Build custom reference model, annotate it, map new data on top
Requirements:
* Reference model file from previous integration effort
* Annotated data object (from previous integration & annotation)
* QC'ed query data

User workflow 3: User loads annotated data with same annotation schema, builds a new reference, maps new data on top
Requirements:
* Annotated datasets with same annotation schema
* Model specification (model class and parameterization)
* QC'ed query data

LuckyMD commented:
Workflows 1 and 2 are more important, workflow 3 has more challenging requirements (only met by CELLxGENE data corpus with cell types mapped to CL)

LuckyMD commented:
Workflow 2 is missing:
* adapt integration modules to output ref models that you can ingest here

LuckyMD commented:
To do:
- [ ] Move workflow 3 to another PR -> unclear whether this workflow is useful
- [ ] Separate PR for workflow 2

LuckyMD commented on src/integrate/scarches/config.vsh.yaml:
```diff
@@ -0,0 +1,140 @@
+          description: "Value used for unlabeled cells"
+          default: "Unknown"
+          required: false
+        - name: "--labels_key"
```
You can get this from the HLCA model directly:
* check latest scvi-tools version for model loading functionality without data
-> can then check required fields via DataRegistry

LuckyMD commented on src/integrate/scarches/config.vsh.yaml:
```diff
@@ -0,0 +1,140 @@
+          type: string
+          description: Name of the reference dataset
+          required: false
+        - name: "--unlabeled_category"
```
Also get this from ref model

LuckyMD commented on src/integrate/scarches/config.vsh.yaml:
```diff
@@ -0,0 +1,140 @@
+          required: false
+        - name: "--predicted_labels_key"
+          type: string
+          description: Name of the column that will contain cell types predicted by scANVI
```
don't use scANVI for this!
XGboost is better

LuckyMD replied:
check code from czi annotation

VladimirShitov replied:
Also, should I simply add code that assumes that we have a classifier? Or is it better to add components for building classifiers and then for running them?

LuckyMD replied:
good point... do we want to separate ref mapping from label transfer? I think mostly this is sth we want to do together, no? If we use an HLCA ref model, we should expect trained classifiers for that already i think.

VladimirShitov replied:
Yes, but what if we want to use several classifiers or if there will be new classifiers available with better quality? With a separate component we can easily apply them to the mapped data. With one component we will need to run mapping each time

LuckyMD replied:
true... either we run all classifiers for one model if we ref map, or we do this separately. I'm okay with both. If it's a separate component we have file I/O to worry about.

LuckyMD commented on src/integrate/scarches/config.vsh.yaml:
```diff
@@ -0,0 +1,140 @@
+        - name: "--input_layer"
+          type: string
+          required: false
+          description: "Input layer to use. If None, X is used"
```
Is this something we can make schema adherent?

LuckyMD replied:
scgen models -> not counts layer
scvi-based models -> counts layer

LuckyMD commented on src/integrate/scarches/config.vsh.yaml:
```diff
@@ -0,0 +1,140 @@
+          type: string
+          default: "sample_id"
+          required: false
+          description: Column name discriminating between your batches.
```
* batch effect key from ref model you can get from the model
* batch effect you want to correct for in query data, you may want to get as input (or just map per batch?... sounds less efficient though)

LuckyMD replied:
give option to encode all query data cells as same batch (maybe when not specified) -> default :)

VladimirShitov replied:
Looks like, scvi-tools always [assumes](https://docs.scvi-tools.org/en/stable/tutorials/notebooks/query_hlca_knn.html#Download-query-data) that batch is stored under "dataset" key
<img width="755" alt="image" src="https://user-images.githubusercontent.com/35199218/203361824-3edc1230-2a9a-4486-84d1-daedb90a612e.png">

VladimirShitov replied:
I added input parameter to set such key if it doesn't exist. But we should also store batch variable under this key (or change scvi-tools code)

LuckyMD commented on src/integrate/scarches/config.vsh.yaml:
```diff
@@ -0,0 +1,140 @@
+        - name: "--var_input"
+          type: string
+          required: false
+          description: ".var column containing highly variable genes. By default, do not subset genes."
```
not needed for workflow 1 and 2

LuckyMD replied:
get the gene set from the model

LuckyMD commented on src/integrate/scarches/config.vsh.yaml:
```diff
@@ -0,0 +1,140 @@
+          default: "X_integrated_scanvi"
+          required: false
+          description: "In which .obsm slot to store the resulting integrated embedding."
+    - name: "Early stopping arguments"
```
do we need this for scArches? I don't think so... so not needed for workflows 1 and 2 ref mapping

LuckyMD replied:
unless needed for fine-tuning?

LuckyMD commented on src/integrate/scarches/config.vsh.yaml:
```diff
@@ -0,0 +1,140 @@
+          default: 0.0
+          description: "Minimum change in the monitored quantity to qualify as an improvement, 
+                        i.e. an absolute change of less than min_delta, will count as no improvement."
+    - name: "Learning parameters"
```
not needed for workflows 1 and 2

LuckyMD replied:
maybe for fine-tuning?

LuckyMD replied:
if needed: defaults from scvi component?

VladimirShitov replied:
Removed these parameters, so they will be default from the scvi_tools

LuckyMD commented on src/integrate/scarches/script.py:
```diff
@@ -0,0 +1,149 @@
+}
+
+
+def map_to_existing_reference(adata_query, model_path, check_val_every_n_epoch=1) -> np.ndarray:
```
maybe make check_val_every_n_epoch a parameter to enable fast test runs

LuckyMD commented on src/integrate/scarches/script.py:
```diff
@@ -0,0 +1,149 @@
+    """
+    print("SCVI-tools version:", scvi.__version__)
+
+    scvi.model.SCANVI.prepare_query_anndata(adata_query, model_path)
```
do we need to specify layers here?

LuckyMD replied:
ensure same layer setup as ref model

LuckyMD commented on src/integrate/scarches/script.py:
```diff
@@ -0,0 +1,149 @@
+    scvi.model.SCANVI.prepare_query_anndata(adata_query, model_path)
+
+    # Load query data into the model
+    vae_query = scvi.model.SCANVI.load_query_data(
```
need to make model specification more flexible for non HLCA models in future
* ocular mega atlas -> scvi

LuckyMD commented on src/integrate/scarches/script.py:
```diff
@@ -0,0 +1,149 @@
+
+    """
+    print("SCVI-tools version:", scvi.__version__)
+
```
Need to load ref model here first to check fields that are required and maybe also model that is required. Then format query data correctly according to ref model

Can we load the model without knowing the model class?

VladimirShitov replied:
> Can we load the model without knowing the model class?
Yes we can, did it
> Need to load ref model here first to check fields that are required and maybe also model that is required. Then format query data correctly according to ref model
Doesn't `prepare_query_anndata()` do it?

LuckyMD replied:
Good point... does it also do this for layer specification though?

LuckyMD commented on src/integrate/scarches/script.py:
```diff
@@ -0,0 +1,149 @@
+
+    scvi.model.SCANVI.convert_legacy_save(f"{directory}/HLCA_reference_model", f"{directory}/HLCA_reference_model_new")
+
+    return f"{directory}/HLCA_reference_model_new"
```
Can this be handled by nextflow to not have to script the model download every time?

@rcannood

VladimirShitov replied:
I actually have a [code](https://github.com/openpipelines-bio/openpipeline/pull/138/files#diff-145f7fc62c5a064abf7ce59aec9a5587e942ad0d9648796a5edbd14cb8a85396) for downloading model only once. Could we run this script and put the model (and a test file) on AWS?

LuckyMD commented on src/integrate/scarches/script.py:
```diff
@@ -0,0 +1,149 @@
+
+        with tempfile.TemporaryDirectory() as directory:
+            model_path = _download_HLCA_reference_model(directory)
+            vae_query, adata_query = map_to_existing_reference(adata_query, model_path=model_path)
```
this is ideally handled by nextflow... somehow ^^

DriesSchaumont replied:
True. One way to do this is to make the `reference` type `file` and set the default to the url (https://zenodo.org/record/6337966/files/HLCA_reference_model.zip). However, this also means that if we want to pass another reference, this also needs to be a path/url. But if we upload them to s3, `s3://` urls are also supported. I think this is a better solution. 

Correct me if im wrong @rcannood

rcannood replied:
I would indeed set the reference to `type: file`. The nextflow module will then https:// and s3:// protocols natively, and will cache the reference model rather than download it during every run.

VladimirShitov replied:
Did it!

LuckyMD commented on src/integrate/scarches/script.py:
```diff
@@ -0,0 +1,149 @@
+    adata_query.obsm[par["obsm_output"]] = vae_query.get_latent_representation()
+
+    logger.info("Predicting labels for SCANVI")
+    adata_query.obs[par["predicted_labels_key"]] = vae_query.predict()
```
Want to use xgboost prediction ideally or other model (knn-classification?)

VladimirShitov replied:
Will add in a separate component as we discussed

LuckyMD reviewed (changes requested):
Move out workflow 3 to other PR... and check model for input obs columns.

DriesSchaumont commented on workflows/resources_test_scripts/scvi_tools_test_data.sh:
```diff
@@ -0,0 +1,55 @@
+  --output "${OUT}_pancreas_ref_test.h5mu"
+
+rm ${OUT}_pancreas_ref_test.h5ad
+rm ${OUT}_pancreas_ref_test.h5ad
```
There is a duplicate line here in the code.

DriesSchaumont commented on src/integrate/scarches/script.py:
```diff
@@ -0,0 +1,149 @@
+}
+
+
+def map_to_existing_reference(adata_query, model_path, check_val_every_n_epoch=1) -> np.ndarray:
```
Seems numpy is not imported:

```python
NameError: name 'np' is not defined
```

VladimirShitov replied:
Thank you! This type hint was also incorrect

DriesSchaumont commented on src/integrate/scvi_tools/script.py:
```diff
@@ -0,0 +1,215 @@
+        adata_query = align_proteins_names(adata_reference, mdata_query, adata_query)
+        
+    # Reorder genes and pad missing genes with 0s
+    model.prepare_query_anndata(adata_query, vae_reference)
```
I am getting an error here when running the tests:

```python3
AttributeError: type object 'SCVI' has no attribute 'prepare_query_anndata'
```

VladimirShitov replied:
Done, should work now

DriesSchaumont reviewed (changes requested):
Thanks for the PR! Looks interesting. Left a few comments for things I observed when running the tests.

rcannood commented:
@VladimirShitov Please let us know when this PR is ready for another review! :)

VladimirShitov commented:
The repository is ready for review. I removed unnecessary details and fixed bugs. Local tests are running fine, but the test pipeline will fail due to the lack of test files. @rcannood , can you please run the script [workflows/resources_test_scripts/hlca_reference_model.sh](https://github.com/openpipelines-bio/openpipeline/pull/138#diff-5ed36765bd1964fd9c983e2f1251c48bf97296ecac3afdb0e03f0bc4f1d347d6) and put the resulting files on AWS so that test script could use them?

DriesSchaumont commented on src/integrate/scarches/config.vsh.yaml:
```diff
@@ -0,0 +1,118 @@
+          alternatives: ["-r"]
+          type: string
+          description: Name of the reference dataset
+          required: false
```
`required` is set to `False`. Could you set the default?

VladimirShitov replied:
It is probably a good idea to set `required` to True instead

DriesSchaumont commented on src/integrate/scarches/config.vsh.yaml:
```diff
@@ -0,0 +1,118 @@
+  namespace: "integrate"
+  version: "dev"
+  description: "Performs reference mapping with scArches"
+  authors:
```
Could use the __inherit__ functionality as described here: #151

DriesSchaumont commented on src/integrate/scarches/script.py:
```diff
@@ -0,0 +1,179 @@
+
+    logger.info("Reassigning modality")
+    mdata_query.mod[par["modality"]] = adata_query
+    try:
```
Could you explain why this happens/is needed?

VladimirShitov replied:
I faced a weird error when trying to reassign modality. This is discussed here: https://github.com/scverse/mudata/issues/27

Briefly, after assigning `adata_query` to the modality of `mdata_query`, described shape of `mdata_query` might be inconsistent with a number of vars and shape of the data. When trying to save such object to a file, one can get a KeyError.

It can be solved by running `mdata_query.update()`. Weirdly, sometimes it still throws a KeyError, but then the shape of the mdata and the problem with saving it to the file is fixed.

VladimirShitov replied:
By the way, according to the comments of mudata developers, reassigning modalities is not expected way to work with mudata. But we use it quite often in this project

VladimirShitov replied:
This is not needed anymore, I decided to create new MuData object instead

DriesSchaumont replied:
Good writeup! I learned a thing or two from the issue.

DriesSchaumont reviewed (changes requested):
Thanks for the PR! Well structured 👍 I left some minor comments/questions.

DriesSchaumont commented:
LGTM! Very nice job @VladimirShitov, thanks!
